include/tbbexec/tbb_thread_pool.hpp

```diff
@@ -0,0 +1,455 @@
+        }
+
+        friend constexpr stdexec::forward_progress_guarantee tag_invoke(stdexec::get_forward_progress_guarantee_t,
+                                                              const DerivedPoolType& pool) noexcept {
+            return pool.forward_progress_guarantee();
+        }
+
+        friend thread_pool_base;
+        explicit scheduler(DerivedPoolType& pool) noexcept : pool_(&pool) {}
+
+        DerivedPoolType* pool_;
+    };
+
+    [[nodiscard]] scheduler get_scheduler() noexcept { return scheduler{static_cast<DerivedPoolType&>(*this)}; }
+
+    /* Is this even needed? Looks like no.
+     * void request_stop() noexcept {
+        // Should this do anything? TBB supports its own flavor of cancelation at a tbb::task_group level
+        //
+    https://spec.oneapi.io/versions/latest/elements/oneTBB/source/task_scheduler/scheduling_controls/task_group_context_cls.html
+        // but not at the tbb::task_arena level.
+        // https://spec.oneapi.io/versions/latest/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.html
+    }*/
+
+    [[nodiscard]] std::uint32_t available_parallelism() const { return static_cast<DerivedPoolType&>(*this).available_parallelism(); }
+
+private:
+    void enqueue(task_base* task, std::uint32_t tid = 0) noexcept { static_cast<DerivedPoolType&>(*this).enqueue(task, tid); }
+    void bulk_enqueue(task_base* task, std::uint32_t n_threads) noexcept {
+        for (std::size_t tid = 0; tid < n_threads; ++tid) {
+            this->enqueue(task, tid);
+        }
+    }
+};
+
+} // namespace detail
+
+
+class tbb_thread_pool : public detail::thread_pool_base<tbb_thread_pool> {
+public:
+    //! Constructor forwards to tbb::task_arena constructor:
+    template <typename... Args>
+    explicit tbb_thread_pool(Args&&... args) : arena_{std::forward<Args>(args)...} {
+        arena_.initialize();
+    }
+
+    [[nodiscard]] std::uint32_t available_parallelism() const { return arena_.max_concurrency(); }
+private:
+    [[nodiscard]] static constexpr stdexec::forward_progress_guarantee forward_progress_guarantee() {
+        return stdexec::forward_progress_guarantee::parallel;
+    }
+
+    friend detail::thread_pool_base<tbb_thread_pool>;
+
+    template <typename PoolType, typename ReceiverId>
+    friend class operation;
+
+    void enqueue(task_base* task, std::uint32_t tid = 0) noexcept {
+        arena_.enqueue([task, tid] { task->__execute(task, /*tid=*/tid); });
+    }
+
+    tbb::task_arena arena_{tbb::task_arena::attach{}};
+};
+
+
+template <typename PoolType, typename ReceiverId>
+class operation : task_base {
+    using Receiver = stdexec::__t<ReceiverId>;
+    friend class detail::thread_pool_base<PoolType>;
+
+    PoolType& pool_;
+    Receiver receiver_;
+
+    explicit operation(PoolType& pool, Receiver&& r) : pool_(pool), receiver_(std::move(r)) {
+        this->__execute = [](task_base* t, std::uint32_t /* tid What is this needed for? */) noexcept {
```

Comment:
@ericniebler Can you comment on the `tid` arg that I'm ignoring here?

Reply:
`tid` is the thread id. It is used in the implementation of `bulk` to assign a chunk of the input to each thread in the pool. It's safe to ignore here.